README
NAME
    ElasticSearch::Transport::Curl - DEPRECATED: A libcurl based HTTP
    backend for ElasticSearch

VERSION
    version 0.08

SYNOPSIS
        use ElasticSearch;
        my $e = ElasticSearch->new(
            servers     => 'search.foo.com:9200',
            transport   => 'curl',
            timeout     => '10',
        );

DESCRIPTION
    THIS MODULE IS DEPRECATED.

    ************************************************************************
    ******

    Because of the name clash between ElasticSearch.pm and Elasticsearch.pm
    the official Perl client is now called: Search::Elasticsearch.

    See https://github.com/elasticsearch/elasticsearch-perl/issues/20 for
    details.

    This distribution will be removed from CPAN in 2015. Please update your
    code.

    ************************************************************************
    ******

    ElasticSearch::Transport::Curl uses WWW::Curl and thus libcurl
    <http://curl.haxx.se/libcurl/> to talk to ElasticSearch over HTTP.

    This is by far the fastest HTTP backend - 60% faster than the next
    fastest ElasticSearch::Transport::HTTPTiny, but does require a C
    compiler.

SEE ALSO
    *   ElasticSearch

    *   ElasticSearch::Transport

    *   ElasticSearch::Transport::HTTPLite

    *   ElasticSearch::Transport::HTTPTiny

    *   ElasticSearch::Transport::Curl

    *   ElasticSearch::Transport::AEHTTP

    *   ElasticSearch::Transport::AECurl

    *   ElasticSearch::Transport::Thrift

AUTHOR
    Clinton Gormley <drtech@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Clinton Gormley.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for ElasticSearch-Transport-Curl

0.08    2013-03-06

        Deprecated the Elasticsearch namespace in favour of Search::Elasticsearch.
        See https://github.com/elasticsearch/elasticsearch-perl/issues/20

0.07    2012-11-14
        Previous change still didn't work. Copying tests from ES to this repo

0.06    2012-11-14
        Previous .tar.gz package was corrupt

0.05    2012-11-07
        Added a max_content_length attribute to ElasticSearch::Transport
        to avoid posting too large a body to elasticsearch. The value
        is auto-detected in versions 0.19.12 onwards

0.04    2012-02-17
        Use ElasticSearch::Transport::code_to_error to centralise HTTP error
        code mapping.
        Now requires ElasticSearch v 0.48

0.03    2011-08-27
        Added support for deflate()

0.02    2011-08-26

        Fixed some dependencies

0.01    2011-08-26

        First release


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601",
    "Test::More" => 0
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "A libcurl based HTTP backend for ElasticSearch",
  "dist_author" => [
    "Clinton Gormley <drtech\@cpan.org>"
  ],
  "dist_name" => "ElasticSearch-Transport-Curl",
  "dist_version" => "0.03",
  "license" => "perl",
  "module_name" => "ElasticSearch::Transport::Curl",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "ElasticSearch" => "0.44",
    "ElasticSearch::Transport" => 0,
    "Encode" => 0,
    "WWW::Curl::Easy" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/ElasticSearch/Transport/Curl.pm
package ElasticSearch::Transport::Curl;

use strict;
use warnings FATAL => 'all';
use WWW::Curl::Easy;
use Encode qw(decode_utf8 encode_utf8);
use ElasticSearch 0.60;
use parent 'ElasticSearch::Transport';

our $VERSION = '0.08';

#===================================
sub protocol     {'http'}
sub default_port {9200}
#===================================

#===================================
sub send_request {
#===================================
    my $self   = shift;
    my $server = shift;
    my $params = shift;

    my $method = $params->{method};
    my $uri = $self->http_uri( $server, $params->{cmd}, $params->{qs} );

    my $client = $self->client;
    $client->setopt( CURLOPT_URL, $uri );

    my $data = $params->{data};
    $data = ''
        if !defined $data
        and $method eq 'POST' || $method eq 'PUT';

    if ( defined $data ) {
        $data = encode_utf8($data);
        $self->check_content_length( \$data );
        $client->setopt( CURLOPT_POSTFIELDS,    $data );
        $client->setopt( CURLOPT_POSTFIELDSIZE, length $data );
    }

    $client->setopt( CURLOPT_CUSTOMREQUEST, $method );
    $client->setopt( CURLOPT_NOBODY,        1 )
        if $method eq 'HEAD';

    $client->setopt( CURLOPT_HTTPHEADER, ['Expect:'] );
    $client->setopt( CURLOPT_ENCODING, 'deflate' )
        if $self->deflate;

    my $content;
    $client->setopt( CURLOPT_WRITEDATA, \$content );

    my $retcode = $client->perform;
    $content = decode_utf8($content)
        if defined $content;

    my ( $code, $msg );
    if ( $retcode == 0 ) {
        $code = $client->getinfo(CURLINFO_RESPONSE_CODE);
        return $content if $code >= 200 && $code <= 209;
        $msg = $self->http_status($code);
    }
    else {
        $msg = $client->strerror($retcode) . " ($retcode) " . $client->errbuf;
        $code = $retcode;
    }

    my $type = $self->code_to_error($code)
        || (
        $code eq '28'
        ? 'Timeout'
        : $code  eq '7'     # can't connect
        || $code eq '55'    # can't write
        || $code eq '56'    # can't read
        || $code eq '18'    # partial file
        || $code eq '25'    # upload failed
        ? 'Connection'
        : 'Request'
        );

    my $error_params = {
        server      => $server,
        status_code => $code,
        status_msg  => $msg,
    };

    if ( $type eq 'Request' or $type eq 'Conflict' or $type eq 'Missing' ) {
        $error_params->{content} = $content;
    }
    $self->throw( $type, $msg . ' (' . $code . ')', $error_params );
}

#===================================
sub client {
#===================================
    my $self = shift;
    unless ( $self->{_client}{$$} ) {
        my $client = WWW::Curl::Easy->new;
        $client->setopt( CURLOPT_TIMEOUT, $self->timeout );
        $client->setopt( CURLOPT_HEADER,  0 );
        $client->setopt( CURLOPT_VERBOSE, 1 );
        $self->{_client} = { $$ => $client };

    }
    my $client = $self->{_client}{$$};
    $client->setopt( CURLOPT_HTTPGET, 1 );
    return $client;
}

# ABSTRACT: DEPRECATED: A libcurl based HTTP backend for ElasticSearch


1;

__END__

=pod

=encoding UTF-8

=head1 NAME

ElasticSearch::Transport::Curl - DEPRECATED: A libcurl based HTTP backend for ElasticSearch

=head1 VERSION

version 0.08

=head1 SYNOPSIS

    use ElasticSearch;
    my $e = ElasticSearch->new(
        servers     => 'search.foo.com:9200',
        transport   => 'curl',
        timeout     => '10',
    );

=head1 DESCRIPTION

THIS MODULE IS DEPRECATED.

******************************************************************************

Because of the name clash between ElasticSearch.pm and Elasticsearch.pm the
official Perl client is now called: Search::Elasticsearch.

See https://github.com/elasticsearch/elasticsearch-perl/issues/20 for details.

This distribution will be removed from CPAN in 2015. Please update your code.

******************************************************************************

ElasticSearch::Transport::Curl uses L<WWW::Curl> and thus
L<libcurl|http://curl.haxx.se/libcurl/> to talk to ElasticSearch
over HTTP.

This is by far the fastest HTTP backend - 60% faster than the next fastest
L<ElasticSearch::Transport::HTTPTiny>, but does require a C compiler.

=head1 SEE ALSO

=over

=item * L<ElasticSearch>

=item * L<ElasticSearch::Transport>

=item * L<ElasticSearch::Transport::HTTPLite>

=item * L<ElasticSearch::Transport::HTTPTiny>

=item * L<ElasticSearch::Transport::Curl>

=item * L<ElasticSearch::Transport::AEHTTP>

=item * L<ElasticSearch::Transport::AECurl>

=item * L<ElasticSearch::Transport::Thrift>

=back

=head1 AUTHOR

Clinton Gormley <drtech@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Clinton Gormley.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.013.
use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "DEPRECATED: A libcurl based HTTP backend for ElasticSearch",
  "AUTHOR" => "Clinton Gormley <drtech\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "ElasticSearch-Transport-Curl",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "ElasticSearch::Transport::Curl",
  "PREREQ_PM" => {
    "ElasticSearch" => "0.60",
    "ElasticSearch::Transport" => 0,
    "Encode" => 0,
    "WWW::Curl::Easy" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ElasticSearch::SearchBuilder" => 0,
    "Test::Differences" => 0,
    "Test::Exception" => 0,
    "Test::More" => "0.96",
    "lib" => 0
  },
  "VERSION" => "0.08",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "ElasticSearch" => "0.60",
  "ElasticSearch::SearchBuilder" => 0,
  "ElasticSearch::Transport" => 0,
  "Encode" => 0,
  "Test::Differences" => 0,
  "Test::Exception" => 0,
  "Test::More" => "0.96",
  "WWW::Curl::Easy" => 0,
  "lib" => 0,
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



